pkg/controller/ingress-shim/helper.go

```diff
@@ -0,0 +1,62 @@
+/*
+Copyright 2020 The Jetstack cert-manager contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+)
+
+var (
+	errNilCertificate           = errors.New("the supplied Certificate pointer was nil")
+	errInvalidIngressAnnotation = errors.New("invalid ingress annotation")
+)
+
+func translateIngressAnnotations(crt *cmapi.Certificate, annotations map[string]string) error {
+	if crt == nil {
+		return errNilCertificate
+	}
+
+	var changes []func()
```

Comment:
This is a very weird approach of doing this...

Reply:
I'm doing it to avoid mutating the supplied certificate if any of the annotations are found to have an error.
It may be overkill, since the calling reconciler will *currently* not attempt to update or create the certificate if this step fails.
Perhaps it'd be better to have a separate `ValidateIngressAnnotations` which runs first.

Reply:
I've simplified this. PTAL

/assign @meyskens